src/Orleans/Core/Grain.cs

```diff
@@ -39,7 +36,7 @@ protected IGrainFactory GrainFactory
         /// </summary>
         protected IServiceProvider ServiceProvider 
         {
-            get { return Runtime?.ServiceProvider; }
+            get { return Data.ServiceProvider ?? Runtime?.ServiceProvider; }
```

Comment:
The latest commits have `IGrainActivationContext` as a scoped service that grains can use to access other scoped services, and eventually bind to grain activation lifecycle events (these are not part of the activation context yet), so that there could be additional infrastructure attached to grains (ie: storage adapters, timers, schedulers, or other infrastructure components that require some lifetime coupling to the grain instance)